### ISSUE TITLE
Uncaught Error: [string "bundle:deps/dgram.lua"]:77: bad argument #2 to 'udp_send' (Invalid IP address or port) when running agent setup

### ISSUE THREAD

**Issue by sunnz** [virgo-agent-toolkit/rackspace-monitoring-agent#835]
Hi,

I got the following error when running rackspace-monitoring-agent --setup on a Rackspace cloud server:

```
Setup Settings:
  Agent ID: ef6ebd09-...
  Config File: /etc/rackspace-monitoring-agent.cfg


Found existing Agent Token for ef6ebd09-...

Writing token to /etc/rackspace-monitoring-agent.cfg: done

Testing Agent connectivity to Cloud Monitoring Service...

Wed Aug 12 10:49:08 2015 INF: Using detected agent ID (id=ef6ebd09-...)
Uncaught Error: [string "bundle:deps/dgram.lua"]:77: bad argument #2 to 'udp_send' (Invalid IP address or port)
stack traceback:
	[C]: in function 'udp_send'
	[string "bundle:deps/dgram.lua"]:77: in function 'send'
	[string "bundle:deps/dns.lua"]:673: in function 'udp_iter'
	[string "bundle:deps/dns.lua"]:682: in function 'resolveSrv'
	[string "bundle:deps/virgo/client/endpoint.lua"]:51: in function 'series'
	[string "bundle:deps/virgo/client/endpoint.lua"]:48: in function 'getHostInfo'
	[string "bundle:deps/virgo/client/connection_stream.lua"]:332: in function 'createConnection'
	[string "bundle:deps/virgo/client/connection_stream.lua"]:97: in function 'iterator'
	[string "bundle:deps/async/init.lua"]:18: in function 'forEach'
	[string "bundle:deps/virgo/client/connection_stream.lua"]:103: in function 'series'
	...
	[string "bundle:deps/virgo/machineidentity.lua"]:56: in function 'done'
	[string "bundle:deps/virgo/machineidentity.lua"]:73: in function 'callback'
	[string "bundle:deps/core.lua"]:183: in function 'handler'
	[string "bundle:deps/core.lua"]:241: in function 'emit'
	[string "bundle:deps/childprocess.lua"]:136: in function <[string "bundle:deps/childprocess.lua"]:133>
	[C]: in function 'run'
	[string "bundle:deps/luvit/init.lua"]:52: in function <[string "bundle:deps/luvit/init.lua"]:47>
	[C]: in function 'xpcall'
	[string "bundle:deps/luvit/init.lua"]:47: in function 'fn'
	[string "bundle:deps/require.lua"]:291: in function <[string "bundle:deps/require.lua"]:256>
```

The os is Linux CentOS 7 and RHEL 7. I have tried both 2.10 from the Rackspace repo and 2.11 from source, both produced the same problems.

However, installing 1.10 from source actually worked on all our rackspace cloud server for now, setup went fine and it is all working properly.

**Updated error with original formatting.** I was on a phone and somehow the phone didn't preserve the format when copying and pasting into github...

**Comment by rphillips** [comment]
Sounds like your resolve.conf is invalid or nonexistant. Is this the case?

**Comment by sunnz** [comment]
Ah, we do have a valid resolv.conf but the problem could be because we are using IPv6 address for the nameserver, for example, `nameserver 2001:4860:4860::8888`.

I will update resolv.conf with a IPv4 address and see what happens.

**Comment by sunnz** [comment]
Nope, just updated resolv.conf with an IPv4 DNS server, still splitting out the same error.

**Comment by rphillips** [comment]
can you make sure it's not a dos file, and it's a unix file

**Comment by sunnz** [comment]
Pretty sure it is not a dos file, all I did was updating it with vim...

**Comment by rphillips** [comment]
is this server connected to the public internet?

**Comment by sunnz** [comment]
Yes, see:

```
$ dig @4.2.2.6 _monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com srv

; <<>> DiG 9.9.4-RedHat-9.9.4-18.el7_1.1 <<>> @4.2.2.6 _monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com srv
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 18757
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 8192
;; QUESTION SECTION:
;_monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com. IN SRV

;; ANSWER SECTION:
_monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com. 245 IN SRV 0 5 443 agent-endpoint-ord.monitoring.api.rackspacecloud.com.

;; Query time: 147 msec
;; SERVER: 4.2.2.6#53(4.2.2.6)
;; WHEN: Wed Aug 12 11:33:53 AEST 2015
;; MSG SIZE  rcvd: 166

$ dig @4.2.2.6 agent-endpoint-ord.monitoring.api.rackspacecloud.com.

; <<>> DiG 9.9.4-RedHat-9.9.4-18.el7_1.1 <<>> @4.2.2.6 agent-endpoint-ord.monitoring.api.rackspacecloud.com.
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33588
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 8192
;; QUESTION SECTION:
;agent-endpoint-ord.monitoring.api.rackspacecloud.com. IN A

;; ANSWER SECTION:
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.14
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.11
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.13
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.18
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.17
agent-endpoint-ord.monitoring.api.rackspacecloud.com. 300 IN A 50.57.61.12

;; Query time: 155 msec
;; SERVER: 4.2.2.6#53(4.2.2.6)
;; WHEN: Wed Aug 12 11:34:07 AEST 2015
;; MSG SIZE  rcvd: 177

$ ping -c 1 agent-endpoint-ord.monitoring.api.rackspacecloud.com.
PING agent-endpoint-ord.monitoring.api.rackspacecloud.com (50.57.61.11) 56(84) bytes of data.
64 bytes from agent-endpoint-ord.monitoring.api.rackspacecloud.com (50.57.61.11): icmp_seq=1 ttl=50 time=230 ms

--- agent-endpoint-ord.monitoring.api.rackspacecloud.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 230.483/230.483/230.483/0.000 ms

$ openssl s_client -connect agent-endpoint-ord.monitoring.api.rackspacecloud.com:443 -servername agent-endpoint-ord.monitoring.api.rackspacecloud.com 
CONNECTED(00000003)
depth=0 C = US, ST = Texas, L = San Antonio, O = "Rackspace US, Inc", OU = Cloud Monitoring, CN = agent-endpoint-ord.monitoring.api.rackspacecloud.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = Texas, L = San Antonio, O = "Rackspace US, Inc", OU = Cloud Monitoring, CN = agent-endpoint-ord.monitoring.api.rackspacecloud.com
verify error:num=27:certificate not trusted
verify return:1
depth=0 C = US, ST = Texas, L = San Antonio, O = "Rackspace US, Inc", OU = Cloud Monitoring, CN = agent-endpoint-ord.monitoring.api.rackspacecloud.com
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/C=US/ST=Texas/L=San Antonio/O=Rackspace US, Inc/OU=Cloud Monitoring/CN=agent-endpoint-ord.monitoring.api.rackspacecloud.com
   i:/C=US/ST=Texas/L=San Antonio/O=Rackspace US, Inc/OU=Cloud Monitoring/CN=monitoring-ca.rackspace.com
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDrzCCApcCAQ4wDQYJKoZIhvcNAQEFBQAwgZAxCzAJBgNVBAYTAlVTMQ4wDAYD
VQQIEwVUZXhhczEUMBIGA1UEBxMLU2FuIEFudG9uaW8xGjAYBgNVBAoTEVJhY2tz
cGFjZSBVUywgSW5jMRkwFwYDVQQLExBDbG91ZCBNb25pdG9yaW5nMSQwIgYDVQQD
Exttb25pdG9yaW5nLWNhLnJhY2tzcGFjZS5jb20wHhcNMTUwNjA1MTUyNDM5WhcN
MTcwNjA0MTUyNDM5WjCBqTELMAkGA1UEBhMCVVMxDjAMBgNVBAgMBVRleGFzMRQw
EgYDVQQHDAtTYW4gQW50b25pbzEaMBgGA1UECgwRUmFja3NwYWNlIFVTLCBJbmMx
GTAXBgNVBAsMEENsb3VkIE1vbml0b3JpbmcxPTA7BgNVBAMMNGFnZW50LWVuZHBv
aW50LW9yZC5tb25pdG9yaW5nLmFwaS5yYWNrc3BhY2VjbG91ZC5jb20wggEiMA0G
CSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDQzBuQ7FR8uYlpN2HAghuXGI5YBVRf
HY+auq8tWMOB8YnHNEYcuu13g9weELxetc+cY19tkCM4Hdgvvb10YkgJ0X0MAh7u
lDwP6eDRwOHVrQbLM78FnbZJfrKeRpyROK7pUd0/q5pL2rSlGS4khPzkkQQ5U/bo
VVH4nfb744ngbIpM2zoUjyLUyvuQlil98h0HYECr89dIBSPOYn21dwcRAniBwaP+
8xTof4J7AY98CErwB+jjNisybw1oCxjFTrsjj/zEiajy5TbWrE2JtHQoySUGhAxD
LGk5yQpZM1qixgz3/veTCa2EqJjR9RQJlsFGjmu7XQILtWSj3HTBzLi/AgMBAAEw
DQYJKoZIhvcNAQEFBQADggEBAMA6xg5dVKyTMwDDaMlRhpszwDPe3f7zxHJ2fIl2
fVn/bwTT029eYJwLoKxlyn6B/vo5mYFaRMr1Y85Impxw/59wMpksjCcdmD61LzYR
OzAtHHY4E9LXiwhfepH7vgWqOk+21wVCHx0SAehGfQ804PG+O3Nt+Q4WqimsGVp1
qrnCjgGYvxd5JXm3J8/Z8wbT1tu8vSrL+q/AdIeYG8Oexv6N0tZQn2nzC9rAVCUA
1dyD/VQQDV4eEC6tv29Qf9dnuSsht6c3RHm6Pbn6hydmQsR4zWa/ULvodV/tsOYu
0yuNoQ0nyySDx2ZO0b/BepOmju6rsr1SSVbfGvDs9CCJLqE=
-----END CERTIFICATE-----
subject=/C=US/ST=Texas/L=San Antonio/O=Rackspace US, Inc/OU=Cloud Monitoring/CN=agent-endpoint-ord.monitoring.api.rackspacecloud.com
issuer=/C=US/ST=Texas/L=San Antonio/O=Rackspace US, Inc/OU=Cloud Monitoring/CN=monitoring-ca.rackspace.com
---
No client certificate CA names sent
Server Temp Key: ECDH, prime256v1, 256 bits
---
SSL handshake has read 1463 bytes and written 436 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES128-GCM-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-GCM-SHA256
    Session-ID: 8B561A1299EA48B0A15B27842F1B21F50DA74EE4804200451CB922448929705B
    Session-ID-ctx: 
    Master-Key: 8A19B3BD699A5D1353D28F2F4CE49FCE8A7761F6D2273C4C110C0670F7606A1954F85516333BCC8B55F9A9ABD7D4E6DE
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1439343311
    Timeout   : 300 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
DONE
```

**Comment by sunnz** [comment]
Also, remember that version 1.1.0 are actually working when I downgraded it on the server:

```
Setup Settings:
  Agent ID: ef6ebd09-...
  Config File: /etc/rackspace-monitoring-agent.cfg
  State Directory: /var/run/rackspace-monitoring-agent/states


Found existing Agent Token for ef6ebd09-...

Writing token to /etc/rackspace-monitoring-agent.cfg: done

Testing Agent connectivity to Cloud Monitoring Service...

Wed Aug 12 02:12:45 2015 INF: Using detected agent ID (id=ef6ebd09-...)
Wed Aug 12 02:12:45 2015 INF: Starting agent ef6ebd09-... (guid=ff69a5fb-..., version=1.1.0-0, bundle_version=1.1.0-0)
Wed Aug 12 02:12:45 2015 INF: Upgrades are disabled
Wed Aug 12 02:12:47 2015 INF: 2001:4801:7902:1:0:a:4323:54:443 (hostname=agent-endpoint-ord.monitoring.api.rackspacecloud.com connID=1) -> Connected
Wed Aug 12 02:12:47 2015 INF: 2001:4801:7902:1:0:a:4323:54:443 (hostname=agent-endpoint-ord.monitoring.api.rackspacecloud.com connID=1) -> Connection has been authenticated to SRV:_monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com
Wed Aug 12 02:12:47 2015 INF: 2001:4800:7902:1:0:a:6039:41:443 (hostname=agent-endpoint-dfw.monitoring.api.rackspacecloud.com connID=1) -> Connected
Wed Aug 12 02:12:47 2015 INF: 2001:4800:7902:1:0:a:6039:41:443 (hostname=agent-endpoint-dfw.monitoring.api.rackspacecloud.com connID=1) -> Connection has been authenticated to SRV:_monitoringagent._tcp.dfw1.prod.monitoring.api.rackspacecloud.com

Agent successfuly connected!

In order to execute checks, the agent must be associated with a Cloud Monitoring Entity.

Wed Aug 12 02:12:48 2015 INF: Confd -> config_file post overall success
Wed Aug 12 02:12:49 2015 INF: 2a00:1a48:7902:1:0:a:603:612:443 (hostname=agent-endpoint-lon.monitoring.api.rackspacecloud.com connID=1) -> Connected
Wed Aug 12 02:12:49 2015 INF: 2a00:1a48:7902:1:0:a:603:612:443 (hostname=agent-endpoint-lon.monitoring.api.rackspacecloud.com connID=1) -> Connection has been authenticated to SRV:_monitoringagent._tcp.lon3.prod.monitoring.api.rackspacecloud.com
Agent already associated Entity with id=... and label=...

Your Agent configuration is now complete.

To start the Agent on your server, now run:

    service rackspace-monitoring-agent start
```

**Comment by rphillips** [comment]
Any chance you could give me access to this server? or a server like it that exhibits the same issue?

My ssh keys are here: https://github.com/rphillips.keys

**Comment by sunnz** [comment]
It is going to be tricky as this is a company server... I might be able to make a sanitised copy.

**Comment by rphillips** [comment]
Hmm. I'm not seeing this with a brand new centos 7 server.

**Comment by rphillips** [comment]
One other thing we can try is moving /etc/resolv.conf to /etc/resolv.conf.backup (to disable it), and try the --setup with the agent again. The newer agent will default to google DNS... It isn't ideal, but will tell us if the problem lies in the resolv.conf parser. Additionally, a copy of the resolv.conf you are using would be helpful.

**Comment by sunnz** [comment]
Disabling resolv.conf now just get:

```
Setup Settings:
  Agent ID: a0cfe2d2-981b-...
  Config File: /etc/rackspace-monitoring-agent.cfg

Error: EAI_NONAME
```
Sure, the resolv.conf file I have right now is simply:

```
nameserver 192.168.5.1
```

This is the one that worked with 1.1.0 in the above (https://github.com/virgo-agent-toolkit/rackspace-monitoring-agent/issues/835#issuecomment-130136915)

I have also tried to put the Rackspace DNS in resolv.conf:

```
nameserver 119.9.60.63
nameserver 119.9.60.62
```

The Rackspace DNS didn't work. They are actually configured as forwarders in 192.168.5.1.

Since you mentioned Google DNS I have just put 8.8.8.8 into resolv.conf, and the setup works now, I am going to try this on other servers to see how they go.

**Comment by sunnz** [comment]
I think I've found the issue, yes it has to do with my resolv.conf, the following does not work for any DNS servers:

```
nameserver 192.168.5.1 # comment
```

It works after removing the comment:

```
nameserver 192.168.5.1 # comment
```

I guess the man page for resolv.conf do say that semicolon (;) or hash character (#) has to be in the first column to be treated as comments, though other stuff and the 1.x version worked? Anyway I will push out an update to resolv.conf to all servers and hopefully they should all work. Thanks for hanging around to help!

**Comment by rphillips** [comment]
Perfect!!! We will patch a few places to make the error better, and it will make it into the next release. Thank you for helping to debug it.

I'll keep this open until we push the changes.

**Comment by sunnz** [comment]
Excellent thanks @rphillips!

I also found that this doesn't work with an IPv6 address even without the comment in resolv.conf, I got a different error though:

```
Could not lookup SRV record for _monitoringagent._tcp.dfw1.prod.monitoring.api.rackspacecloud.com
```

Do you mind if I open a new issue, or should I keep it in this thread?

**Comment by rphillips** [comment]
let's just keep this one open... what is the ipv6 address? It could be the ipv6 address in an abbreviated form that we do not support.

**Comment by sunnz** [comment]
Sure. This is what I get by spinning up a new CentOS 7 server for testing only:

```
# cat /etc/resolv.conf
nameserver 2001:4860:4860::8888

# sudo rackspace-monitoring-agent --setup --username s... --apikey 3...

Setup Settings:
  Agent ID: 132a423e-...
  Config File: /etc/rackspace-monitoring-agent.cfg


Found existing Agent Token for 132a423e-...

Writing token to /etc/rackspace-monitoring-agent.cfg: done

Testing Agent connectivity to Cloud Monitoring Service...

Wed Aug 12 04:09:12 2015 INF: Using detected agent ID (id=132a423e-...)
Wed Aug 12 04:09:12 2015 INF: Starting agent 132a423e-... (guid=fda6ff47-..., version=2.0.1, bundle_version=2.1.10)
Wed Aug 12 04:09:12 2015 INF: Upgrades are disabled
Wed Aug 12 04:09:20 2015 ERR: Could not lookup SRV record for _monitoringagent._tcp.dfw1.prod.monitoring.api.rackspacecloud.com
Wed Aug 12 04:09:20 2015 ERR: nil -> Error resolving (Maximum attempts reached)
Wed Aug 12 04:09:20 2015 INF: SRV:_monitoringagent._tcp.dfw1.prod.monitoring.api.rackspacecloud.com -> Retrying connection in 73279ms
Wed Aug 12 04:09:20 2015 ERR: Could not lookup SRV record for _monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com
Wed Aug 12 04:09:20 2015 ERR: nil -> Error resolving (Maximum attempts reached)
Wed Aug 12 04:09:20 2015 INF: SRV:_monitoringagent._tcp.ord1.prod.monitoring.api.rackspacecloud.com -> Retrying connection in 77161ms
Wed Aug 12 04:09:20 2015 ERR: Could not lookup SRV record for _monitoringagent._tcp.lon3.prod.monitoring.api.rackspacecloud.com
Wed Aug 12 04:09:20 2015 ERR: nil -> Error resolving (Maximum attempts reached)
Wed Aug 12 04:09:20 2015 INF: SRV:_monitoringagent._tcp.lon3.prod.monitoring.api.rackspacecloud.com -> Retrying connection in 62278ms
^C
```

We used to use something like `fdcc:eeff:2014:811:be76:4ffe:f18e:b31` when we were running agent 1.1.0, which is not working with 2.10 neither.

**Comment by rphillips** [comment]
Found the issue [1]. I will get a new release out either late tonight (CST) or early tomorrow. Thank you for the report.

[1] https://github.com/luvit/luvit/commit/37536697415566bfb7cf96617bd462b245e2b753

**Comment by rphillips** [comment]
I released 1.2.11 with the fix, and it is working for me. Thank you for the help!

**Comment by sunnz** [comment]
Thanks @rphillips for the prompt update, all our servers are back to using IPv6 DNS server happily with 2.1.11 now!
